crates/autopilot/src/solvable_orders.rs

```diff
@@ -340,72 +337,32 @@ impl SolvableOrdersCache {
-        // expired/invalid orders in the SQL query and basically can return the whole
-        // table when filters with default values are used.
-        let (db_solvable_orders, previous_creation_timestamp) = {
-            let cache_data = {
-                let lock = self.cache.lock().await;
-                match &*lock {
-                    Some(cache)
-                        if cache.last_order_creation_timestamp
-                            > INITIAL_ORDER_CREATION_TIMESTAMP =>
-                    {
-                        Some((
-                            cache.solvable_orders.orders.clone(),
-                            cache.last_order_creation_timestamp,
-                            cache.solvable_orders.latest_settlement_block,
-                        ))
-                    }
-                    _ => None,
-                }
-            };
-
-            let min_valid_to = now_in_epoch_seconds()
-                + u32::try_from(self.min_order_validity_period.as_secs())
-                    .context("min_order_validity_period is not u32")?;
-            match cache_data {
-                Some((current_orders, last_order_creation_timestamp, latest_settlement_block)) => (
-                    self.persistence
-                        .solvable_orders_after(
-                            current_orders,
-                            last_order_creation_timestamp,
-                            latest_settlement_block,
-                            min_valid_to,
-                        )
-                        .await?,
-                    last_order_creation_timestamp,
-                ),
-                None => (
-                    self.persistence.all_solvable_orders(min_valid_to).await?,
-                    INITIAL_ORDER_CREATION_TIMESTAMP,
-                ),
+    /// Returns currently solvable orders.
+    async fn get_solvable_orders(&self) -> Result<SolvableOrders> {
+        let min_valid_to = now_in_epoch_seconds()
+            + u32::try_from(self.min_order_validity_period.as_secs())
+                .context("min_order_validity_period is not u32")?;
+
+        let lock = self.cache.lock().await;
+        let solvable_orders = match &*lock {
+            Some(cache) => {
+                // Only use incremental query after cache already got initialized
+                // because it's not optimized for very long durations.
+                let fetch_orders = self.persistence.solvable_orders_after(
+                    cache.solvable_orders.orders.clone(),
+                    cache.solvable_orders.fetched_from_db,
+                    cache.solvable_orders.latest_settlement_block,
+                    min_valid_to,
+                );
+                drop(lock);
+                fetch_orders.await?
+            }
+            _ => {
+                drop(lock);
+                self.persistence.all_solvable_orders(min_valid_to).await?
```

Comment:
Maybe we can avoid manual `drop` calls by returning `solvable_orders_future` instead and calling `.await?` below this section to drop the lock automatically.

Reply:
Good idea!